### PR TITLE
[Ballerina 1.2.x] Add XML code coverage report generation to testerina

### LINKED ISSUE #28972: [Testerina] Add XML code coverage report generation to ballerina 1.2.x
**Description:**

- [x] Introduce includes flag
- [x] Add module wise coverage XML
- [x] Add fix to consider partially covered lines as covered

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Dilhasha**
## Purpose
Add XML code coverage report generation to testerina

Fixes #28972

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 1]
```diff
@@ -124,4 +132,52 @@ private void createReport(final IBundleCoverage bundleCoverage) {
...
+                        execFileLoader.getExecutionDataStore().getContents());
+
+                visitor.visitBundle(bundleCoverage, null);
+
+                visitor.visitEnd();
+            }
+        }
+    }
+
+    private Collection<ISourceFileCoverage> modifySourceFiles(Collection<ISourceFileCoverage> sourcefiles) {
```
Comment (pramodya1994) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:159]:
```suggestion
    private Collection<ISourceFileCoverage> modifySourceFiles(Collection<ISourceFileCoverage> sourceFiles) {
```

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 2]
```diff
@@ -124,4 +132,52 @@ private void createReport(final IBundleCoverage bundleCoverage) {
...
+        Collection<ISourceFileCoverage> modifiedSourceFiles = new ArrayList<>();
+        for (ISourceFileCoverage sourcefile : sourcefiles) {
```
Comment (pramodya1994) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:161]:
```suggestion
        for (ISourceFileCoverage sourceFile : sourcefiles) {
```

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/PartialCoverageModifiedSourceFile.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    public int getLastLine() {
+        return oldSourceFile.getLastLine();
+    }
+
+    /**
+     * Returns the modified lines instead of lines stored in the original source file.
+     */
+    @Override
+    public ILine getLine(int lineNumber) {
+        if (modifiedLines.size() == 0 || lineNumber < getFirstLine() || lineNumber > getLastLine()) {
```
Comment (pramodya1994) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/PartialCoverageModifiedSourceFile.java:57]: Shall we use `isEmpty()` here?

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/PartialCoverageModifiedSourceFile.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ * Represents a source file containing lines modified to consider partially covered coverage info as fully covered.
+ *
+ * @since 2.0.0
+ */
+public class PartialCoverageModifiedSourceFile implements ISourceFileCoverage {
+
+    private final ISourceFileCoverage oldSourceFile;
+    private final List<ILine> modifiedLines;
+
+    public PartialCoverageModifiedSourceFile(ISourceFileCoverage oldSourcefile, List<ILine> modifiedLines) {
```
Comment (pramodya1994) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/PartialCoverageModifiedSourceFile.java:37]:
```suggestion
    public PartialCoverageModifiedSourceFile(ISourceFileCoverage oldSourceFile, List<ILine> modifiedLines) {
```

Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/PartialCoverageModifiedSourceFile.java]: fixed with f16e2a2

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/PartialCoverageModifiedSourceFile.java (hunk 3 above)
Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/PartialCoverageModifiedSourceFile.java]: fixed with f16e2a2

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 2 above)
Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java]: fixed with f16e2a2

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 1 above)
Comment (Dilhasha) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java]: fixed with f16e2a2

**Review by azinneera — APPROVED** [review]

**Comment by Dilhasha** [comment]
This implementation has changed.

**Pull request closed without merging**